Merged from sdl-tabs https://github.com/Novvum/graphql-playground.git

rajinwonderland opened the pull request:
# Feature Request: SchemaExplorer Tab
---

## Screenshot
**Dark Mode:**
<img width="1440" alt="darkmode" src="https://user-images.githubusercontent.com/15880596/49395272-e041f800-f6ea-11e8-9e25-18ffcbeec24a.png">

**Light Mode:**
<img width="1440" alt="lightmode" src="https://user-images.githubusercontent.com/15880596/49395286-e8019c80-f6ea-11e8-95d2-e5bdf4434a59.png">


## Changes proposed in this pull request:

- Creating an ExplorerTab System for Docs, Schema, and any other new features that may be implemented into the Explorer side tab.

- Renaming Current Schema Tab (DocsExplorer Component) to Docs

- Adding a Schema Tab that shows the SDL version (Using CodeMirror Component) of the schema

- Options to Download Schema in either `.json` or `.graphql` format within the Schema tab

### Proposed File Structure Changes:

```
src
├── ...
├── components
│   ├── ...
│   ├── Playground
│   │   ├── ...
│   │   ├── DocExplorer //  Directory Containing all DocExplorer Components
│   │   │   ├── ...
│   │   ├── ExplorerTabs // Directory Containing all Tab System Components used by the Side Drawer
│   │   │   ├── SideTab.tsx // The Tab Button Component
│   │   │   └── SideTabs.tsx // The TabContainer Module
│   │   ├── SchemaExplorer // Directory Containing all SchemaExplorer Components
│   │   │   ├── SDLEditor.tsx // CodeMirror editor version of the SDL
│   │   │   ├── SDLTypes // DocExplorer styled version of the SDL
│   │   │   │   ├── SDLDocType.tsx 
│   │   │   │   ├── SDLFieldDoc.tsx
│   │   │   │   ├── SDLStyles.tsx
│   │   │   │   ├── SDLType.tsx
│   │   │   │   └── SDLUnionType.tsx
│   │   │   └── SDLView.tsx // SDL View Component
│   │   ├── ...
│   │   └── util
│   │       ├── ...
│   │       ├── createSDL.ts // Utils for SchemaExplorer Components
│   │       └── ...
│   ├── ...
└── ...
```

Todos:

- [ ] Review Design
- [ ] Refresh Endpoint Icon Placement
- [ ] Autopolling Schema from endpoint

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/prisma/graphql-playground?pullRequest=897) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/prisma/graphql-playground?pullRequest=897) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/prisma/graphql-playground?pullRequest=897) it.</sub>

Huvik commented:
Great job! 🚀

rajinwonderland commented:
## Implemented Disabled Comments in SDLView


**Schema View Screenshot** `schema.disabledComments: true`

<img width="1442" alt="screen shot 2018-12-03 at 4 19 12 pm" src="https://user-images.githubusercontent.com/15880596/49409870-53f9fa00-f717-11e8-86c7-2b7a81103f57.png">

**Settings Screenshot**

<img width="1440" alt="screen shot 2018-12-03 at 4 22 39 pm" src="https://user-images.githubusercontent.com/15880596/49409971-b2bf7380-f717-11e8-954e-dac43db149f2.png">


Changes to `ISettings Interface` on [README.md](https://github.com/Novvum/gql-playground/blob/schema-view/README.md) and [src/types.ts](https://github.com/Novvum/gql-playground/blob/schema-view/packages/graphql-playground-react/src/types.ts)

```ts
interface ISettings {
	'general.betaUpdates': boolean
	'editor.theme': Theme
	'editor.reuseHeaders': boolean
	'tracing.hideTracingResponse': boolean
	'editor.fontSize': number
	'editor.fontFamily': string
	'request.credentials': string
	'schema.disableComments': boolean // defaults to true
}
```

_On_ `src/state/workspace/reducers.ts` the `defaultSettings` has the added attribute of `schema.disableComments`

```ts
... const defaultSettings: ISettings = {
  'general.betaUpdates': false,
  'editor.cursorShape': 'line',
  'editor.fontSize': 14,
  'editor.fontFamily': `'Source Code Pro', 'Consolas', 'Inconsolata', 'Droid Sans Mono', 'Monaco', monospace`,
  'editor.theme': 'dark',
  'editor.reuseHeaders': true,
  'prettier.printWidth': 80,
  'request.credentials': 'omit',
  'tracing.hideTracingResponse': true,
  'schema.disableComments': true
}
```

kuldar commented:
Hey Raj, thanks for the PR, looking awesome! 🙌 I checked it out and played around with it a little bit locally - few notes I took:

1. It seems like the left side padding is bit too narrow at the moment - could we increase it by ~20px?
<img width="352" alt="padding" src="https://user-images.githubusercontent.com/1710629/49462830-5df32a00-f7f7-11e8-80e6-f95874bcfce8.png">

----

2. Would it prehaps be better DX if we `overflow:scroll` the schema, rather than linebreaking? It’s easier to scroll sideways to read more data, than it is to decypher the beginnings and ends of linebreaked code.
<img width="340" alt="screenshot 2018-12-04 at 19 07 20" src="https://user-images.githubusercontent.com/1710629/49463073-dbb73580-f7f7-11e8-8e8d-9cb0b197a7e0.png">

----

3. For the sidebar, let's maybe have a consistent color for both the code and header background. The Playground currently has a lot of boxes and areas as it is, so it helps to reduce the visual noise just a bit. For overflow indicator, we could add a slight dropshadow.
<img width="543" alt="screenshot 2018-12-04 at 19 09 43" src="https://user-images.githubusercontent.com/1710629/49463236-38b2eb80-f7f8-11e8-8961-cecb7dfa1395.png">

----

4. Could we also double-check the download button - ideally the height should be the same as the left-hand action buttons. (~30px) and the color bit more toned back.

5. As for the dropdown itself, I realized I hadn't really clearified the different states, so I mocked a better version to illustrate how the hover states should look like. Could we also have it close on blur/outsideClick?
<img width="654" alt="dropdowns" src="https://user-images.githubusercontent.com/1710629/49463443-aced8f00-f7f8-11e8-973b-6d5d80993ddd.png">

----

6. The last one, I promise 😄 Can we tweak the sizing of the vertical buttons just a bit. (By making the "Docs" button ~20px narrower for example).
<img width="74" alt="screenshot 2018-12-04 at 19 17 28" src="https://user-images.githubusercontent.com/1710629/49463699-62b8dd80-f7f9-11e8-8f14-5d5dc8863813.png">

----

As for showing/hiding comments - this feels more like a setting than an action. (I can’t imagine people toggling between the states too often). Should we just keep the option in the settings, or do we feel strongly about having a separate UI for that?

(Btw changing the schema comments setting didn't affect my schema when trying out locally for some reason).

rajinwonderland commented:
Perfect! Thanks @kuldar! Will start working on those specs!

rajinwonderland commented:
## Style Changes & Tweaks

As requested by @kuldar 

**Comments**
- Separated `SDLHeader` component into it's own file (originally in SDLTypes/SDLStyles.tsx)

**Checklist**
- [x]  Increase Editor Padding by 24px
- [x]  Disable Line Breaking on Editor: *Allow for* `overflow scroll`
- [x]  SDLHeader style changes
    - [x]  Add `box-shadow` as an overflow indicator
        - `box-shadow: 0px 1px 3px rgba(17, 17, 17, 0.1);`
    - [x]  Fix background color so it matches the editor
- [x]  Fix Height on Download & Option Buttons (32px)
- [x]  Fix Color on Download & Option Buttons (On Figma)
- [x]  Fix Stylings for different states on Download & Option Buttons
- [x]  Close Options on blur/outside click
- [x]  Look into Settings `schema.disabledComments` on local server
- [x]  Decrease width of Docs by ~20px *Change fixed width of 60px to 8px padding*
- [x]  Change copy from 'Search the schema...' to 'Search the docs...'

rajinwonderland commented:
@kuldar for some reason, the cmd+save button isn't working on the Settings editor. I think this may be a deeper issue. Try the Save button to disable the comments for now and let me know if it still persists! 

**Tab Screenshots:** 📷 📸 

Theme | With Comment | W/O Comments | Options Button
------------ | ------------- | ----------- | ---------
Light |  <img width="125" alt="lwc" src="https://user-images.githubusercontent.com/15880596/49486615-02736d00-f7f4-11e8-983f-46870ab738a3.png"> | <img width="125" alt="lw oc" src="https://user-images.githubusercontent.com/15880596/49486622-0bfcd500-f7f4-11e8-9798-644eb03138c2.png"> | <img width="125" alt="lwc options" src="https://user-images.githubusercontent.com/15880596/49486628-15863d00-f7f4-11e8-9dd1-fc03407d1655.png">
Dark | <img width="125" alt="dwc" src="https://user-images.githubusercontent.com/15880596/49486655-3189de80-f7f4-11e8-9631-d8ef5004fba5.png"> | <img width="125" alt="dw oc" src="https://user-images.githubusercontent.com/15880596/49486669-39e21980-f7f4-11e8-9ca5-e46a1871ba7e.png"> | <img width="125" alt="dw oc options" src="https://user-images.githubusercontent.com/15880596/49486684-423a5480-f7f4-11e8-9dd8-20c5a04603a1.png">

Docs Screenshot 📜 📓 📸 

<img width="250" alt="docs" src="https://user-images.githubusercontent.com/15880596/49486700-5c743280-f7f4-11e8-8590-b7a1fa097d8b.png">

rajinwonderland commented:
Hey @kuldar agreed. I'll go ahead and see what I can do to fix that!

rajinwonderland commented:
Hey @kuldar, just added a `tabWidth` prop to the side tabs based on your figma designs.

**Screenshots** 📸

_Dark Theme_ 

<img width="332" alt="screen shot 2018-12-05 at 11 46 19 am" src="https://user-images.githubusercontent.com/15880596/49539827-67c26f00-f883-11e8-9731-a9891946114c.png">

_Light Theme_

<img width="332" alt="screen shot 2018-12-05 at 11 47 21 am" src="https://user-images.githubusercontent.com/15880596/49539903-93ddf000-f883-11e8-8bca-287c7c506ef7.png">

kuldar commented:
Awesome, looks good to me! 👍

schickling commented:
A bit more feedback from my side:

- [ ] The spacing for the collapsed tabs seems off

![image](https://user-images.githubusercontent.com/1567498/49600594-d8e94b80-f983-11e8-8924-bf6fc67fe63f.png)

- [ ] Let's add additional line-breaks after each item

```
# change from ...
type AggregateUser {
  count: Int!
}
type BatchPayload {
  count: Long!
}
scalar Long

# to ...
type AggregateUser {
  count: Int!
}

type BatchPayload {
  count: Long!
}

scalar Long
```

rajinwonderland commented:
Hey @schickling thanks for the feedback!

Just pushed some changes

- [X] Fix Tab Spacing in collapsed state
- [X] Add additional line-breaks after each 

## Tab Spacing Fix

<img width="177" alt="tabspacing" src="https://user-images.githubusercontent.com/15880596/49611497-8fced280-f956-11e8-94e9-fb73bad29113.png">

- Changed `right: -2px` to `right: 0px` on SideTab.tsx styled `Tab` component

## Line Breaking on Schema

With Comments  |  Without Comments
-------  | ------
<img width="300" alt="withcomments" src="https://user-images.githubusercontent.com/15880596/49611537-b0972800-f956-11e8-96ea-0c8a6859fefa.png"> | <img width="300" alt="withoutcomments" src="https://user-images.githubusercontent.com/15880596/49611552-b8ef6300-f956-11e8-831e-62c27cec7f50.png">

- Added `addLineBreaks(sdl: string, commentsDisabled?: boolean)` function on createSDL.ts

## Note on Comment Linebreaking

I've opted to removed added newlines on comments since the newlines from each comment, but I'm not sure what would be best.

![new lines on inline comments](https://user-images.githubusercontent.com/15880596/49612237-e50be380-f958-11e8-9e7f-1796bcaa7456.png)

schickling commented:
This looks great. Thanks a lot for the updates. I agree, let's remove the line breaks between fields.

rajinwonderland commented:
Cool! Line breaks are removed on the fields! 😊
